Extend the existing SSE support to enable Customer (provided) Keys SSE-C

rawipfel opened the pull request:
This resolves issue #79 (Support AWS KMS and SSE-C). Example usage:
```
s3.ssea.name=AES256
# base64encode("32_byte_string_needed_for_AES256")
s3.sse.customer.key=MzJfYnl0ZV9zdHJpbmdfbmVlZGVkX2Zvcl9BRVMyNTY=
```

rawipfel commented:
Hi, this built OK for me:
```
[INFO] Building zip: /Users/rwipfel/git/KafkaConnect/kafka-connect-storage-cloud/kafka-connect-s3/target/components/packages/confluentinc-kafka-connect-s3-4.1.1.zip
[INFO]
[INFO] --- maven-failsafe-plugin:2.21.0:integration-test (integration-test) @ kafka-connect-s3 ---
[INFO]
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 0, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO]
[INFO] --- maven-failsafe-plugin:2.21.0:verify (verify) @ kafka-connect-s3 ---
[INFO]
[INFO] --- maven-install-plugin:2.5.2:install (default-install) @ kafka-connect-s3 ---
[INFO] Installing /Users/rwipfel/git/KafkaConnect/kafka-connect-storage-cloud/kafka-connect-s3/target/components/kafka-connect-s3-4.1.1.jar to /Users/rwipfel/.m2/repository/io/confluent/kafka-connect-s3/4.1.1/kafka-connect-s3-4.1.1.jar
[INFO] Installing /Users/rwipfel/git/KafkaConnect/kafka-connect-storage-cloud/kafka-connect-s3/pom.xml to /Users/rwipfel/.m2/repository/io/confluent/kafka-connect-s3/4.1.1/kafka-connect-s3-4.1.1.pom
[INFO] Installing /Users/rwipfel/git/KafkaConnect/kafka-connect-storage-cloud/kafka-connect-s3/target/kafka-connect-s3-4.1.1-tests.jar to /Users/rwipfel/.m2/repository/io/confluent/kafka-connect-s3/4.1.1/kafka-connect-s3-4.1.1-tests.jar
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] kafka-connect-storage-cloud 4.1.1 .................. SUCCESS [  2.263 s]
[INFO] kafka-connect-s3 4.1.1 ............................. SUCCESS [01:04 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:07 min
[INFO] Finished at: 2018-06-02T10:47:28-06:00
[INFO] ------------------------------------------------------------------------
```

rawipfel commented:
/cc @kkonstantine, and @jklukas thanks in advance for reviewing

jklukas commented on kafka-connect-s3/src/main/java/io/confluent/connect/s3/S3SinkConnectorConfig.java:
```diff
@@ -221,6 +224,18 @@ public static ConfigDef newConfigDef() {
+          Type.STRING,
+          SSE_KEY_DEFAULT,
+          Importance.LOW,
+          "The S3 Server Side Encryption Customer Key.",
```
I'd recommend sticking as closely as possible to Amazon's terminology here. I'd refer to this as "customer-provided server-side encryption key (SSE-C)"

rawipfel replied:
Thanks, will fix

rawipfel replied:
I checked the docs and it looks like all of these strings parrot the HTTP header strings: https://docs.aws.amazon.com/AmazonS3/latest/dev/ServerSideEncryptionCustomerKeys.html. I could change them all to be customer (document) friendly, what do you think @kkonstantine?

jklukas commented on kafka-connect-s3/src/main/java/io/confluent/connect/s3/storage/S3OutputStream.java:
```diff
@@ -196,7 +199,9 @@ private MultipartUpload newMultipartUpload() throws IOException {
     ).withCannedACL(cannedAcl);
 
-    if (SSEAlgorithm.KMS.toString().equalsIgnoreCase(ssea)
+    if (StringUtils.isNotBlank(ssea) && StringUtils.isNotBlank(sseKey)) {
```
The `StringUtils.isNotBlank(ssea)` check here seems weak. Should this be `SSEAlgorithm.AES256.toString().equalsIgnoreCase(ssea)`? I'm not sure exactly what the SDK requires here.

rawipfel replied:
Thanks, good point, I will double check but remember reading that only AES256 is supported, I'll make this a stronger check.

rawipfel replied:
Yes, only AES256 is supported: https://docs.aws.amazon.com/AmazonS3/latest/dev/ServerSideEncryptionCustomerKeys.html

rawipfel commented:
Status as of 6/11:

- Tested with Minio S3 (https://github.com/minio/cookbook/blob/master/docs/how-to-use-minio-server-side-encryption-with-aws-cli.md)
- Tested a multi-part upload app with and without the new settings
```
s3.ssea.name=AES256
s3.sse.key=MzJfYnl0ZV9zdHJpbmdfbmVlZGVkX2Zvcl9BRVMyNTY=
```
Addressed both of @jklukas's review comments - thanks for that!

kkonstantine commented on kafka-connect-s3/src/main/java/io/confluent/connect/s3/storage/S3OutputStream.java:
```diff
@@ -71,6 +74,8 @@ public S3OutputStream(String key, S3SinkConnectorConfig conf, AmazonS3 s3) {
     this.key = key;
     this.ssea = conf.getSsea();
+    this.sseKey = conf.getSseKey();
+    this.sseCustomerKey = null;
```
We have all we need to know in the constructor to create the key object or not. 
Thus, I'd suggest making this variable `final` and setting it here as: 
`this.sseCustomerKey = (condition below) ? new key : null;`

rawipfel commented:
Hi @kkonstantine 

Thanks again, I've addressed all of your review comments, and retested on a 5.0.x based branch
```
	s3.sse.customer.key = [hidden]
	s3.sse.kms.key.id =
	s3.ssea.name = AES256
```
Should I open a new PR from: https://github.com/rawipfel/kafka-connect-storage-cloud/tree/SSE-C-5.0.x?

kkonstantine commented:
@rawipfel would you mind rebasing this branch and force pushing it, to keep working on this PR?

You may change target branch on github by clicking on `Edit` at the upper right corner and selecting `5.0.x` instead of `master`

Thanks for addressing the comments!

rawipfel commented:
Ok, I'll look at doing a rebase, this branch SSE-C started from 4.1.1-post (6cb3983). 
My other branch SSE-C-5.0.x started from your 5.0.x (5330dee). Will it be easier to take a PR from that? while I check the rebase option...

kkonstantine commented:
@rawipfel the changes are not that many. Feel free to open an new PR if you prefer.

rawipfel commented:
@kkonstantine yes you're right, not many changes: I did the rebase, cherry-picked today's commits, and retargeted the PR to 5.0.x.

rawipfel commented on kafka-connect-s3/src/main/java/io/confluent/connect/s3/storage/S3OutputStream.java:
```diff
@@ -53,6 +54,8 @@
   private final String bucket;
   private final String key;
   private final String ssea;
+  private final String sseCustomerKeyConfig;
```
Thanks @kkonstantine fixing this now...

rawipfel commented:
Thanks @kkonstantine it was great to work with you and @jklukas, I appreciate the help!

kkonstantine commented:
Also, thanks @jklukas for the initial review!
